File: Step 04 Binary Search/Step 4.2 Applying BS on 2D Array/Row with max 1s/Row with max 1s.cpp
// User function template for C++
class Solution {
  public:
     //optimal approach tc->o(arr.size() * log(2,(arr[i].size())) sc->o(1)

int first_occurence_index(vector<int>&arr,int n,int x){
        int low=0;int high=n-1;
        while(low<=high){
            int mid=(low+high)/2;
            if(arr[mid]>=x){
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return low;
    }

 int rowWithMax1s(vector<vector<int>> arr) {

int cnt_max=0;int row=-1;
for(int i=0;i<arr.size();i++){
    int cnt_ones=arr[i].size()-first_occurence_index(arr[i],arr[i].size(),1);
    if(cnt_ones>cnt_max){
        cnt_max=cnt_ones;
        row=i;
    }
    }
   return row; 
 }
};

File: Step 04 Binary Search/Step 4.2 Applying BS on 2D Array/Row with max 1s/Row with max 1s.java


// User function Template for Java

class Solution {
    public int rowWithMax1s(int matrix[][]) {
        // code here
        int row =  matrix.length;
        int col = matrix[0].length;
        int  maxCount = 0;
        int resInd = -1;
        for(int i=0; i<row; ++i) {
            int low = 0;
            int high = col - 1;
            while(low <= high) {
                int mid = low + (high - low) / 2;
                if(matrix[i][mid] == 1) high = mid - 1;
                else low = mid + 1;
            }
            // if (high < 0) high = 0;
            int count = col - (high + 1);
            if(maxCount < count) {
                maxCount = Math.max(maxCount, count);
                resInd = i;
            }
        }
        return resInd;
    }
}